Add AtomisticsMaintenance

pmrv opened the pull request:
Also add a method that corrects energy_pot of Vasp jobs.

This only adds the maintenance methods, the output parser is not yet updated.

pmrv committed:
Add AtomisticsMaintenance

Also add a method that corrects energy_pot of Vasp jobs

samwaseda commented on pyiron_atomistics/vasp/base.py:
```diff
@@ -53,6 +53,17 @@
+    if self.project_hdf5.get("HDF_VERSION", "0.1.0") == "0.1.0":
+        energy_free = np.array([e[-1] for e in self.project_hdf5["output/generic/dft/scf_energy_pot"]])
+        energy_pot = self.project_hdf5["output/generic/energy_pot"]
+        return not np.allclose(energy_free, energy_pot)
```
Isn't it other way around? Shouldn't it be `True` when they are all close?

pmrv replied:
Oh, it needs to be scf_energy_free.

samwaseda replied:
ah ok now I understand your intention. I thought it was `energy_zero` XD

pmrv committed:
Check against correct energy and import from correct location

pmrv committed:
Remove comment

pmrv committed:
Make scf_energy_kin an array

This fixes a bug where output/generic/energy_pot of VASP AIMD runs have
nonsensical kinetic energy

pmrv committed:
Parse free energy as generic energy instead of zero extrapolated one

pmrv committed:
Save kinetic energy also under energy_kin instead of scf_energy_kin

The latter implies that it is a quantity that changes during SCF
iteration, but it does not, since it is only updated once per ionic
step.

pmrv committed:
Add method to recollect output for correct kinetic energies

pmrv committed:
Add docstrings for update methods

pmrv committed:
Small typo fixes

pmrv committed:
Reload job before re collecting

pmrv commented:
@ahmedabdelkawy and I also found two bugs in the way AIMD runs were parsed that are quite severe:

1. only the final kinetic energy was parsed
2. this was then subtracted from the total electronic energy instead of added

Bug fixes and another update method are included in this PR.

@ahmedabdelkawy I've tested this on a quick MD run, but would be great if you could also try it on your jobs.

pyiron-runner committed:
Format black

pmrv committed:
Fix typo

pmrv committed:
Emit warning during energy_pot fix if energy_kin fix needs to be applied

pmrv commented:
Test failure seems to be related to importing Vasp runs from file, will have to check tomorrow.

pmrv committed:
More typo

pmrv committed:
Pass on kinetic energy only if found

pmrv committed:
Adjust table test

Change the numerical values to the free energy derived ones and use AlmostEqual for float comparisons like a grown up.

ahmedabdelkawy commented on pyiron_atomistics/vasp/base.py:
```diff
@@ -2051,15 +2073,33 @@ def collect(self, directory=os.getcwd(), sorted_indices=None):
+                    log_dict["energy_pot"]
+                    + self.vp_new.vasprun_dict["kinetic_energies"]
                 )
             else:
```
This will never be true, because the if condition only checks whether kinetic_energies key exists or not and not if there are actual values in it. This becomes a problem for static calculations where self.vp_new.vasprun_dict["kinetic_energies"] is an empty list but the key is still there. This will result in log_dict["energy_tot"] being empty.

pmrv replied:
Correct!  I pushed a fix.

ahmedabdelkawy replied:
parsing energy now works perfectly!

pyiron-runner committed:
Format black

pmrv committed:
Actually pick the free energies in case of Vasp>5

ahmedabdelkawy commented:
For misfortunate future generations who will have to work on this again: 

Parsing energies from vasp to pyiron now works as follows:
1. Parsing is done from vasprun.xml and not the outcar
2. Parsing of energy values is done by the following order of files: vasprun.xml -> vasprun.py -> base.py
3. There is a bug in vasprun.xml for vasp 5.xx versions described here: https://ww.vasp.at/forum/viewtopic.php?p=19242 and was taken into consideration
4. Parsed energy terminology is the following:
 * Energy minimization jobs:
  1. job.output.energy_pot  =  free energy (including smearing if used to be consistent with forces)
  2. job.output.energy_tot = job.output.energy_pot

 * MD jobs:
  1. job.output.energy_pot  =  free energy (including smearing if used to be consistent with forces)
  2. job.output.energy_tot = job.output.energy_pot  + job['output/generic/dft/energy_kin'] (kinetic energies)


More accessible quantities that are stored as a dictionary:
- arrays with MxN shape: M = number of ionic steps, N = number of electronic steps (ragged: differs from one ionic step to the other)
1. job['output/generic/dft/scf_energy_free'] -> free energy
2. job['output/generic/dft/scf_energy_int'] -> internal energy, aka energy when the entropy (eg: due to the degeneracy effect of the bands) is equal to zero
3. job['output/generic/dft/scf_energy_kin'] -> kinetic energies
5.  job['output/generic/dft/scf_energy_zero'] -> energy when the electronic temperature (aka sigma) is extrapolated to zero

- same as above but only at the last electronic step for each ionic step:
1. job['output/generic/dft/energy_free'] -> free energy
2. job['output/generic/dft/energy_int'] -> internal energy, aka energy when the entropy (eg: due to the degeneracy effect of the bands) is equal to zero
3. job['output/generic/dft/energy_kin'] -> kinetic energies
4.  job['output/generic/dft/energy_zero'] -> energy when the electronic temperature (aka sigma) is extrapolated to zero

pmrv commented:
@niklassiemer Please bump the minor version for the next release.
